70_FLIPFluids/app_resources/compute/diffusion.comp.hlsl

```diff
@@ -53,90 +56,191 @@ void setAxisCellMaterial(uint32_t3 ID : SV_DispatchThreadID)
     uint3 cmAxisTypes = 0;
     setCellMaterial(cmAxisTypes, cellAxisType);
 
-    axisCellMaterialOutBuffer[cid].xyz = cmAxisTypes;
+    axisCellMaterialOutBuffer[cellIdx].xyz = cmAxisTypes;
 }
 
-[numthreads(WorkgroupSize, 1, 1)]
+[numthreads(WorkgroupGridDim, WorkgroupGridDim, WorkgroupGridDim)]
 void setNeighborAxisCellMaterial(uint32_t3 ID : SV_DispatchThreadID)
 {
-    uint cid = ID.x;
-    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+    int3 cellIdx = ID;
 
     uint3 axisCm = (uint3)0;
-    uint3 this_axiscm = getCellMaterial(axisCellMaterialInBuffer[cid].xyz);
+    uint3 this_axiscm = getCellMaterial(axisCellMaterialInBuffer[cellIdx].xyz);
     setCellMaterial(axisCm, this_axiscm);
 
-    uint3 xp_axiscm = cellIdx.x == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(-1, 0, 0), gridData.gridSize)].xyz);
+    uint3 xp_axiscm = cellIdx.x == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[clampToGrid(cellIdx + int3(-1, 0, 0), gridData.gridSize)].xyz);
     setXPrevMaterial(axisCm, xp_axiscm);
 
-    uint3 xn_axiscm = cellIdx.x == gridData.gridSize.x - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(1, 0, 0), gridData.gridSize)].xyz);
+    uint3 xn_axiscm = cellIdx.x == gridData.gridSize.x - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[clampToGrid(cellIdx + int3(1, 0, 0), gridData.gridSize)].xyz);
     setXNextMaterial(axisCm, xn_axiscm);
```

Comment:
could probably be sped up by preloading a patch into shared memory and doing all neighbour checks and sets locally